client/client.go

```diff
@@ -188,6 +214,284 @@ func (q Query) AddResponse(resp *http.Response) {
+
+	for {
+		_, message, err := c.ReadMessage()
+		if err != nil {
+			if websocket.IsUnexpectedCloseError(err, websocket.CloseNormalClosure) {
+				query.CheckErr(err)
+			}
+			return
+		}
+		answers = append(answers, string(message))
+	}
+}
+
+// GetGRPCBridgeReqBody returns the body of the HTTP request using the
+// HTTP/1.1-gRPC bridge format as described in the Envoy docs
+// https://www.envoyproxy.io/docs/envoy/v1.9.0/configuration/http_filters/grpc_http1_bridge_filter
+func GetGRPCBridgeReqBody() (*bytes.Buffer, error) {
+	// Protocol:
+	// 	. 1 byte of zero (not compressed).
+	// 	. network order (big-endian) of proto message length.
+	// 	. serialized proto message.
+	buf := &bytes.Buffer{}
+	if err := binary.Write(buf, binary.BigEndian, uint8(0)); err != nil {
+		log.Printf("error when packing first byte: %v", err)
+		return nil, err
+	}
+
+	m := &grpc_echo_pb.EchoRequest{}
+	m.Data = "foo"
+
+	pbuf := &proto.Buffer{}
+	if err := pbuf.Marshal(m); err != nil {
+		log.Printf("error when serializing the gRPC message: %v", err)
+		return nil, err
+	}
+
+	if err := binary.Write(buf, binary.BigEndian, uint32(len(pbuf.Bytes()))); err != nil {
+		log.Printf("error when packing message length: %v", err)
+		return nil, err
+	}
+
+	for i := 0; i < len(pbuf.Bytes()); i++ {
+		if err := binary.Write(buf, binary.BigEndian, uint8(pbuf.Bytes()[i])); err != nil {
+			log.Printf("error when packing message: %v", err)
+			return nil, err
+		}
+	}
+
+	return buf, nil
+}
+
+// CallRealGRPC handles real gRPC queries, i.e. queries that use the normal gRPC
+// generated code and the normal HTTP/2-based transport.
+func CallRealGRPC(query Query) {
+	qURL, err := url.Parse(query.URL())
+	if query.CheckErr(err) {
+		log.Printf("grpc url parse failed: %v", err)
+		return
+	}
+
+	const requiredPath = "/echo.EchoService/Echo"
+	if qURL.Path != requiredPath {
+		query.Result()["error"] = fmt.Sprintf("GRPC path %s is not %s", qURL.Path, requiredPath)
+		return
+	}
+
+	if !strings.Contains(qURL.Host, ":") {
+		query.Result()["error"] = fmt.Sprintf("GRPC URL %s has no port", qURL.Host)
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	// Dial runs in the background and thus always appears to succeed. If you
+	// pass grpc.WithBlock() to make it wait for a connection, failures just hit
+	// the deadline rather than returning a useful error like "no such host" or
+	// "connection refused" or whatever. Perhaps they are considered "transient"
+	// and there's some retry logic we need to turn off. Anyhow, we don't pass
+	// grpc.WithBlock(), instead letting the error happen at the request below.
+	// This makes useful error messages visible in most cases.
+	var dialOptions []grpc.DialOption
+	if qURL.Scheme != "https" {
+		dialOptions = append(dialOptions, grpc.WithInsecure())
+	}
+	conn, err := grpc.DialContext(ctx, qURL.Host, dialOptions...)
+	if query.CheckErr(err) {
+		log.Printf("grpc dial failed: %v", err)
```

Comment:
👍

Reply:
Yup, I added this log statement because I was really confused about why it would "time out" connecting to localhostXXX rather than giving me a "host not found" error. This is explained in the comment above. Again, this helped me with debugging.